tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/mhlo_shape_calc_placement.cc

```diff
@@ -0,0 +1,539 @@
+#include <map>
+#include <set>
+#include <string>
+#include <unordered_map>
+
+#include "absl/memory/memory.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // TF:llvm-project
+#include "mlir/Dialect/Tensor/IR/Tensor.h"    // from @llvm-project
+#include "mlir/IR/MLIRContext.h"              // TF:llvm-project
+#include "mlir/Pass/Pass.h"                   // TF:local_config_mlir
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_disc_ops.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/utils/hlo_utils.h"
+
+namespace mlir {
+
+using hlo::kInputPlacementAttr;
+using hlo::kPlaceTyAttr;
+using hlo::kTypeDevice;
+using hlo::kTypeHost;
+using hlo::PlacementType;
+
+namespace mhlo {
+namespace {
+
+bool isTargetDialect(llvm::StringRef dialect_name) {
+  return (dialect_name == "mhlo");
+}
+
+// This pass explicitly place the shape calculating hlo_ops on host side
+// by adding an Attr. Nested FuncOps should be taken into consideration.
+// 1, Normally, the type of kPlaceTyAttr is StringAttr;
+// 2, In case the result type of an hlo op is TupleType, for example TupleOp
+//    or TopKOp, the type of kPlaceTyAttr is an ArrayAttr made of
+//    StringAttr.
+struct PlaceShapeCalcOnHost
+    : public PassWrapper<PlaceShapeCalcOnHost, OperationPass<ModuleOp>> {
+ public:
+  PlaceShapeCalcOnHost() = default;
+  PlaceShapeCalcOnHost(const PlaceShapeCalcOnHost& o){};
+
+  StringRef getArgument() const final { return "place-shape-calc-on-host"; }
+
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<mhlo::MhloDialect, mhlo_disc::MhloDiscDialect>();
+  }
+
+  void runOnOperation() override {
+    // Phase 1: Place the shape calculation subgraph to Host
+    placeShapeCalcSubgraphOnHost();
+    // Phase 2: Place any mhlo ops that calculates I32 on Host
+    placeI32OpsOnHost();
+
+    // TODO(disc): Phase 3: Add placement attribute for TupleOp and
+    // GetTupleElementOp.
+
+    // Phase 4: Insert h2d and d2h OP on cross device edges. Host is CPU. Device
+    // is GPU or CPU
+    addMemcpyNodes();
+  };
+
+ private:
+  // Place the shape calculation subgraph to Host
+  void placeShapeCalcSubgraphOnHost();
+  // Place any mhlo ops that calculates I32 on Host
+  void placeI32OpsOnHost();
+  // Insert h2d and d2h OP on cross device edges.
+  void addMemcpyNodes();
+
+  // for rule based placement strategy, the placement of the op in the list
+  // is up to the placement of the dominant operand
+  const std::unordered_map<std::string, /*dominant operand index*/ int>
+      kPlaceRuleMap = {{"mhlo.dynamic_gather", /*operand*/ 0},
+                       {"mhlo.gather", /*operand*/ 0}};
+
+  const std::unordered_map<std::string, std::set<int>> kShapeCalcOperandMap = {
```

Comment:
Please document the map.
Also this may not be the most efficient data-structure.

Instead of using strings as keys, can we use TypeID? This'll be basically a pointer key unique for the class, and you can use `op->getAbstractOperation()->typeID` as a key.

For the value, I'd use a vector instead of a set, the set really does not bring much here.

Reply:
(Here is the comment I was looking for)